### PR TITLE
Add shell support for win checkvm module and `get_processes` method in `post/common.rb`

### PR DESCRIPTION
**Pull request by pingport80**
## Summary
This PR adds shell session support to `windows_checkvm` module. However there is one limitation for shell sessions over meterpreter. Shell sessions don't have `get_processes` method to check for specific processes to identify if it's a vm. But most checks are based on registry keys which work on both.

## Verification Steps
```
msf6 > handler -H 192.168.29.232 -P 4444 -p windows/x64/shell_reverse_tcp
[*] Payload handler running as background job 7.

[*] Started reverse TCP handler on 192.168.29.232:4444 
msf6 > [*] Command shell session 1 opened (192.168.29.232:4444 -> 192.168.29.232:44208)

msf6 > use post/windows/gather/checkvm 
msf6 post(windows/gather/checkvm) > set session 1 
session => 1
msf6 post(windows/gather/checkvm) > run

[*] Checking if the target is a Virtual Machine ...
[+] This is a VirtualBox Virtual Machine
[*] Post module execution completed
msf6 post(windows/gather/checkvm) > 
```

### CODE REVIEW

**Comment by smcintyre-r7** [comment]
Instead of removing the check when the session is a Meterpreter, how about adding a mixin method that works for both shell sessions and Meterpreter sessions that returns the running processes? You can use `tasklist` in a shell on Windows and parse the results easily enough and on a Meterpreter session you can use the existing `get_processes` method.

The Meterpreter `get_processes` methods will return a hash with additional keys that the shell session won't be able to get from `tasklist`. You can make this consistent by just returning the PID and image name which is probably all that's needed in most cases. If the module author needs the PPID and other pieces of information, they'll need to only support Meterpreter.

Eventually, we could add Linux shell support by analyzing `proc` or using `ps` but for now if you just want to raise a [`NotImplementedError`](https://ruby-doc.org/core-2.5.0/NotImplementedError.html) and focus on Windows that'd be fine too. I think it makes sense to put this into [`lib/msf/core/post/common.rb`](https://github.com/rapid7/metasploit-framework/blob/31eab90c74d9ba5301d9939d6205c1c67d141f6d/lib/msf/core/post/common.rb#L36) where we already have `has_pid?`. The common mixin would be a better place than the process mixin because that one is pretty Meterpreter-specific.

We should avoid sacrificing accuracy though for flexibility, especially when we can get the process list easily enough.

**Comment by pingport80** [comment]
Okay I'll add `get_processes` to the `common.rb` :+1:

File: lib/msf/core/post/common.rb [hunk 1]
```diff
@@ -251,4 +251,31 @@ def command_exists?(cmd)
...
+      tasklist = cmd_exec('tasklist').split("\n")
+      tasklist.each do |p|
+        properties = p.split
+        process = {}
+        process['name'] = properties[0].to_s
+        process['pid'] = properties[1].to_s
+        processes.push(process)
+      end
+    else
+      ps_aux = cmd_exec('ps aux').split("\n")
```
Comment (smcintyre-r7) [inline lib/msf/core/post/common.rb:269]: Looks like this one should filter out the first line.

```
msf6 post(windows/gather/checkvm) > irb
[*] Starting IRB shell...
[*] You are in post/windows/gather/checkvm

>> get_processes
=> 
[{"name"=>"COMMAND", "pid"=>"PID"},
 {"name"=>"/usr/lib/systemd/systemd", "pid"=>"1"},
 {"name"=>"kthreadd", "pid"=>"2"},
```

**Review by smcintyre-r7 — CHANGES_REQUESTED** [review]
Can you also please update `#has_pid?` to use this implementation of `get_processes`? That would ensure that the two functions yield consistent results.

Something like this should work:
```ruby
def has_pid?(pid)
  get_processes.any? { |proc| proc['pid'] == pid }
end
```

File: lib/msf/core/post/common.rb [hunk 2]
```diff
@@ -251,4 +251,31 @@ def command_exists?(cmd)
...
+      return session.sys.process.get_processes
+    end
+    processes = []
+    if session.platform == 'windows'
+      tasklist = cmd_exec('tasklist').split("\n")
+      tasklist.each do |p|
+        properties = p.split
+        process = {}
+        process['name'] = properties[0].to_s
+        process['pid'] = properties[1].to_s
```
[the remark this answers is not in the document]
Comment (pingport80) [reply lib/msf/core/post/common.rb:265]: okay, I'll change that. :+1:

File: lib/msf/core/post/common.rb [hunk 3]
```diff
@@ -251,4 +251,31 @@ def command_exists?(cmd)
...
+    if session.type == 'meterpreter'
+      return session.sys.process.get_processes
+    end
+    processes = []
+    if session.platform == 'windows'
+      tasklist = cmd_exec('tasklist').split("\n")
+      tasklist.each do |p|
+        properties = p.split
+        process = {}
+        process['name'] = properties[0].to_s
```
[the remark this answers is not in the document]
Comment (pingport80) [reply lib/msf/core/post/common.rb:264]: okay :+1:

File: lib/msf/core/post/common.rb [hunk 4]
```diff
@@ -251,4 +251,31 @@ def command_exists?(cmd)
     raise "Unable to check if command `#{cmd}' exists"
   end
 
+  def get_processes
+    if session.type == 'meterpreter'
+      return session.sys.process.get_processes
+    end
+    processes = []
+    if session.platform == 'windows'
+      tasklist = cmd_exec('tasklist').split("\n")
```
[the remark this answers is not in the document]
Comment (pingport80) [reply lib/msf/core/post/common.rb:260]: Okay!

File: lib/msf/core/post/common.rb [hunk 5]
```diff
@@ -251,4 +251,31 @@ def command_exists?(cmd)
     raise "Unable to check if command `#{cmd}' exists"
   end
 
+  def get_processes
+    if session.type == 'meterpreter'
+      return session.sys.process.get_processes
```
[the remark this answers is not in the document]
Comment (pingport80) [reply lib/msf/core/post/common.rb:256]: yeah, I'll change that.

**Comment by pingport80** [comment]
> Can you also please update `#has_pid?`

Sure!

File: lib/msf/core/post/common.rb (hunk 1 above)
Comment (pingport80) [reply lib/msf/core/post/common.rb:269]: yeah, thanks I'll change that.

File: lib/msf/core/post/common.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (pingport80) [reply lib/msf/core/post/common.rb:256]: Removed the unwanted keys. :+1:

File: lib/msf/core/post/common.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (pingport80) [reply lib/msf/core/post/common.rb:264]: Removed `to_s` from both the blocks.

File: lib/msf/core/post/common.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (pingport80) [reply lib/msf/core/post/common.rb:265]: Added conversion to int

**Review by smcintyre-r7 — APPROVED** [review]
I just pushed up some minor changes in 4f921e46f35d1f921712453a333b70f4533ff1b7 that filter the output before processing it, prepend the static system entry and renamed it for consistency with Meterpreter. The module is looking great now, I also tested the `get_processes` function with a Linux shell session where it looked consistent with Mettle. The only inconsistency I noticed is that sometimes the full binary path is returned as the name and other times it's not but that seems pretty minor.

<details>
<summary>Testing Output</summary>

```
msf6 exploit(windows/smb/psexec) > exploit

[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] 192.168.159.129:445 - Connecting to the server...
[*] 192.168.159.129:445 - Authenticating to 192.168.159.129:445 as user 'smcintyre'...
[*] 192.168.159.129:445 - Selecting PowerShell target
[*] 192.168.159.129:445 - Executing the payload...
[*] Sending stage (336 bytes) to 192.168.159.129
[+] 192.168.159.129:445 - Service start timed out, OK if running a command or non-service executable...
[*] Command shell session 1 opened (192.168.159.128:4444 -> 192.168.159.129:63800) at 2021-05-04 16:28:01 -0400

^Z
Background session 1? [y/N]  y
msf6 exploit(windows/smb/psexec) > previous 
[-] There isn't a previous module at the moment
msf6 exploit(windows/smb/psexec) > use post/windows/gather/checkvm 
msf6 post(windows/gather/checkvm) > irb
[*] Starting IRB shell...
[*] You are in post/windows/gather/checkvm

>> get_processes
=> 
[{"name"=>"[System Process]", "pid"=>0},
 {"name"=>"System", "pid"=>4},
 {"name"=>"Registry", "pid"=>92},
 {"name"=>"smss.exe", "pid"=>336},
 {"name"=>"csrss.exe", "pid"=>424},
 {"name"=>"wininit.exe", "pid"=>500},
 {"name"=>"csrss.exe", "pid"=>508},
 {"name"=>"winlogon.exe", "pid"=>588},
 {"name"=>"services.exe", "pid"=>636},
 {"name"=>"lsass.exe", "pid"=>648},
 {"name"=>"svchost.exe", "pid"=>760},
 {"name"=>"fontdrvhost.exe", "pid"=>768},
 {"name"=>"fontdrvhost.exe", "pid"=>776},
 {"name"=>"svchost.exe", "pid"=>880},
 {"name"=>"svchost.exe", "pid"=>932},
 {"name"=>"dwm.exe", "pid"=>992},
 {"name"=>"svchost.exe", "pid"=>632},
 {"name"=>"svchost.exe", "pid"=>696},
 {"name"=>"svchost.exe", "pid"=>872},
 {"name"=>"svchost.exe", "pid"=>384},
 {"name"=>"svchost.exe", "pid"=>1052},
 {"name"=>"svchost.exe", "pid"=>1120},
 {"name"=>"svchost.exe", "pid"=>1172},
 {"name"=>"svchost.exe", "pid"=>1292},
 {"name"=>"svchost.exe", "pid"=>1332},
 {"name"=>"svchost.exe", "pid"=>1368},
 {"name"=>"svchost.exe", "pid"=>1412},
 {"name"=>"svchost.exe", "pid"=>1472},
 {"name"=>"svchost.exe", "pid"=>1496},
 {"name"=>"svchost.exe", "pid"=>1532},
 {"name"=>"svchost.exe", "pid"=>1584},
 {"name"=>"svchost.exe", "pid"=>1744},
 {"name"=>"svchost.exe", "pid"=>1816},
 {"name"=>"svchost.exe", "pid"=>1824},
 {"name"=>"svchost.exe", "pid"=>1848},
 {"name"=>"Memory", "pid"=>0},
 {"name"=>"svchost.exe", "pid"=>1940},
 {"name"=>"svchost.exe", "pid"=>1960},
 {"name"=>"svchost.exe", "pid"=>1968},
 {"name"=>"svchost.exe", "pid"=>2008},
 {"name"=>"svchost.exe", "pid"=>2020},
 {"name"=>"svchost.exe", "pid"=>2080},
 {"name"=>"svchost.exe", "pid"=>2232},
 {"name"=>"svchost.exe", "pid"=>2240},
 {"name"=>"svchost.exe", "pid"=>2252},
 {"name"=>"svchost.exe", "pid"=>2308},
 {"name"=>"svchost.exe", "pid"=>2344},
 {"name"=>"spoolsv.exe", "pid"=>2456},
 {"name"=>"svchost.exe", "pid"=>2528},
 {"name"=>"svchost.exe", "pid"=>2596},
 {"name"=>"svchost.exe", "pid"=>2652},
 {"name"=>"sihost.exe", "pid"=>2740},
 {"name"=>"svchost.exe", "pid"=>2768},
 {"name"=>"svchost.exe", "pid"=>2844},
 {"name"=>"svchost.exe", "pid"=>2960},
 {"name"=>"taskhostw.exe", "pid"=>2996},
 {"name"=>"svchost.exe", "pid"=>2912},
 {"name"=>"svchost.exe", "pid"=>3096},
 {"name"=>"svchost.exe", "pid"=>3108},
 {"name"=>"svchost.exe", "pid"=>3120},
 {"name"=>"svchost.exe", "pid"=>3164},
 {"name"=>"IpOverUsbSvc.exe", "pid"=>3192},
 {"name"=>"svchost.exe", "pid"=>3236},
 {"name"=>"svchost.exe", "pid"=>3296},
 {"name"=>"VGAuthService.exe", "pid"=>3352},
 {"name"=>"vmtoolsd.exe", "pid"=>3360},
 {"name"=>"svchost.exe", "pid"=>3372},
 {"name"=>"ctfmon.exe", "pid"=>3380},
 {"name"=>"MsMpEng.exe", "pid"=>3388},
 {"name"=>"svchost.exe", "pid"=>3432},
 {"name"=>"svchost.exe", "pid"=>3620},
 {"name"=>"svchost.exe", "pid"=>3660},
 {"name"=>"svchost.exe", "pid"=>3852},
 {"name"=>"svchost.exe", "pid"=>4008},
 {"name"=>"dllhost.exe", "pid"=>4144},
 {"name"=>"WmiPrvSE.exe", "pid"=>4448},
 {"name"=>"explorer.exe", "pid"=>4660},
 {"name"=>"svchost.exe", "pid"=>4872},
 {"name"=>"SearchIndexer.exe", "pid"=>5012},
 {"name"=>"svchost.exe", "pid"=>5096},
 {"name"=>"svchost.exe", "pid"=>3844},
 {"name"=>"SettingSyncHost.exe", "pid"=>3500},
 {"name"=>"powershell.exe", "pid"=>5176},
 {"name"=>"conhost.exe", "pid"=>5188},
 {"name"=>"msdtc.exe", "pid"=>5568},
 {"name"=>"StartMenuExperienceHost.e", "pid"=>5848},
 {"name"=>"RuntimeBroker.exe", "pid"=>5920},
 {"name"=>"SearchApp.exe", "pid"=>6016},
 {"name"=>"RuntimeBroker.exe", "pid"=>536},
 {"name"=>"YourPhone.exe", "pid"=>6224},
 {"name"=>"LockApp.exe", "pid"=>6836},
 {"name"=>"ShellExperienceHost.exe", "pid"=>6852},
 {"name"=>"RuntimeBroker.exe", "pid"=>6988},
 {"name"=>"RuntimeBroker.exe", "pid"=>7092},
 {"name"=>"RuntimeBroker.exe", "pid"=>6764},
 {"name"=>"svchost.exe", "pid"=>7648},
 {"name"=>"GoogleCrashHandler.exe", "pid"=>7876},
 {"name"=>"SecurityHealthSystray.exe", "pid"=>7968},
 {"name"=>"SecurityHealthService.exe", "pid"=>8000},
 {"name"=>"GoogleCrashHandler64.exe", "pid"=>8052},
 {"name"=>"vm3dservice.exe", "pid"=>8104},
 {"name"=>"RuntimeBroker.exe", "pid"=>8176},
 {"name"=>"vmtoolsd.exe", "pid"=>6680},
 {"name"=>"OneDrive.exe", "pid"=>7464},
 {"name"=>"ApplicationFrameHost.exe", "pid"=>7832},
 {"name"=>"WinStore.App.exe", "pid"=>8156},
 {"name"=>"RuntimeBroker.exe", "pid"=>1672},
 {"name"=>"svchost.exe", "pid"=>4396},
 {"name"=>"svchost.exe", "pid"=>1864},
 {"name"=>"svchost.exe", "pid"=>2028},
 {"name"=>"svchost.exe", "pid"=>7132},
 {"name"=>"SgrmBroker.exe", "pid"=>5984},
 {"name"=>"svchost.exe", "pid"=>3032},
 {"name"=>"svchost.exe", "pid"=>7440},
 {"name"=>"MoUsoCoreWorker.exe", "pid"=>2548},
 {"name"=>"svchost.exe", "pid"=>6996},
 {"name"=>"svchost.exe", "pid"=>7500},
 {"name"=>"TextInputHost.exe", "pid"=>8592},
 {"name"=>"dllhost.exe", "pid"=>8956},
 {"name"=>"SecHealthUI.exe", "pid"=>9112},
 {"name"=>"SecurityHealthHost.exe", "pid"=>9144},
 {"name"=>"svchost.exe", "pid"=>8396},
 {"name"=>"svchost.exe", "pid"=>8344},
 {"name"=>"SecurityHealthHost.exe", "pid"=>8228},
 {"name"=>"powershell.exe", "pid"=>752},
 {"name"=>"powershell.exe", "pid"=>7616},
 {"name"=>"conhost.exe", "pid"=>7660},
 {"name"=>"Microsoft.Photos.exe", "pid"=>6032},
 {"name"=>"RuntimeBroker.exe", "pid"=>4440},
 {"name"=>"powershell.exe", "pid"=>5340},
 {"name"=>"conhost.exe", "pid"=>2060},
 {"name"=>"powershell.exe", "pid"=>7076},
 {"name"=>"conhost.exe", "pid"=>6672},
 {"name"=>"powershell.exe", "pid"=>3928},
 {"name"=>"conhost.exe", "pid"=>2316},
 {"name"=>"svchost.exe", "pid"=>4848},
 {"name"=>"svchost.exe", "pid"=>7584},
 {"name"=>"DbgX.Shell.exe", "pid"=>8252},
 {"name"=>"EngHost.exe", "pid"=>6268},
 {"name"=>"svchost.exe", "pid"=>7984},
 {"name"=>"devenv.exe", "pid"=>8796},
 {"name"=>"PerfWatson2.exe", "pid"=>2980},
 {"name"=>"Microsoft.ServiceHub.Cont", "pid"=>8940},
 {"name"=>"ServiceHub.ThreadedWaitDi", "pid"=>2100},
 {"name"=>"ServiceHub.VSDetouredHost", "pid"=>1988},
 {"name"=>"ServiceHub.IdentityHost.e", "pid"=>492},
 {"name"=>"ServiceHub.Host.CLR.x86.e", "pid"=>1420},
 {"name"=>"ServiceHub.SettingsHost.e", "pid"=>4048},
 {"name"=>"vcpkgsrv.exe", "pid"=>4684},
 {"name"=>"ServiceHub.Host.CLR.x86.e", "pid"=>5952},
 {"name"=>"ServiceHub.TestWindowStor", "pid"=>600},
 {"name"=>"Microsoft.Alm.Shared.Remo", "pid"=>3776},
 {"name"=>"vcpkgsrv.exe", "pid"=>6752},
 {"name"=>"powershell.exe", "pid"=>6524},
 {"name"=>"conhost.exe", "pid"=>8920},
 {"name"=>"powershell.exe", "pid"=>5144},
 {"name"=>"conhost.exe", "pid"=>7128},
 {"name"=>"svchost.exe", "pid"=>5708},
 {"name"=>"powershell.exe", "pid"=>1700},
 {"name"=>"conhost.exe", "pid"=>1128},
 {"name"=>"cmd.exe", "pid"=>1776},
 {"name"=>"tasklist.exe", "pid"=>8668}]
>> exit
msf6 post(windows/gather/checkvm) > run

[*] Checking if the target is a Virtual Machine ...
[+] This is a VMware Virtual Machine
[*] Post module execution completed
msf6 post(windows/gather/checkvm) >
```
</details>

Once the unit tests pass, I'll get this landed.

**Comment by smcintyre-r7** [comment]
@msjenkins-r7 test this please.

**Comment by smcintyre-r7** [comment]
# Release Notes
This adds shell session support to the `post/windows/gather/checkvm` module. This also notably adds cross-platform support for getting a list of running processes using shell and Meterpreter sessions.

**Comment by pingport80** [comment]
Thanks @zeroSteiner for reviewing and merging it. :+1:

**Comment by jmartin-r7** [comment]
@pingport80 sorry about the revert here.  We identified some issues with dependency cycles and reverted in #15244.  

I am thinking the process methods in common may be good to factor out into `msf/core/post/process.rb` which could be dependent on `Msf::Post::File` without creating a cycle.  This would move the `has_pid?` method as well so I may take pass at this and offer as a PR towards your branch that shows how I am thinking this might work.

**Comment by pingport80** [comment]
Okay makes sense! :slightly_smiling_face:

**Pull request merged**